classes/task/course_backup_s3_task.php

```diff
@@ -0,0 +1,168 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_lcbackupcoursestep\task;
+
+use backup_plan_dbops;
+use tool_lcbackupcoursestep\s3\helper;
+use tool_lifecycle\settings_type;
+use tool_lifecycle\local\manager\settings_manager;
+
+/**
+ * Defines task for course backup.
+ *
+ * @package     tool_lcbackupcoursestep
+ * @copyright   2024 Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class course_backup_s3_task extends \core\task\adhoc_task {
+    /**
+     * Return the component name.
+     */
+    public function get_component() {
+        return 'tool_lcbackupcoursestep';
+    }
+
+    /**
+     * Run the adhoc task and preform the backup.
+     */
+    public function execute() {
+        global $DB;
+        $customdata = $this->get_custom_data();
+
+        // Workflow id.
+        $workflowid = $customdata->workflowid;
+
+        // Process ID of a lifecycle process.
+        $processid = $customdata->processid;
+
+        // Instance ID of the step.
+        $stepinstanceid = $customdata->stepinstanceid;
+
+        // Course ID to backup.
+        $courseid = $customdata->courseid;
+
+        // Check the IDs.
+        // Check if the course exists.
+        if (!$courseid || !$course = $DB->get_record('course', ['id' => $courseid])) {
+            mtrace('Invalid course id: ' . $courseid . ', task aborted.');
+            return;
+        }
+        // Check if the workflow exists.
+        if (!$workflowid || !$DB->record_exists('tool_lifecycle_workflow', ['id' => $workflowid])) {
+            mtrace('Invalid workflow id: ' . $workflowid . ', task aborted.');
+            return;
+        }
+
+        // Check if the backup task is already running on the same course.
+        $lockfactory = \core\lock\lock_config::get_lock_factory('course_backup_adhoc');
```

Comment:
why do we need this? Would it be managed by core as we schedule the task with "$checkforexisting" option?

Reply:
If we have 2 workflows with this step, they may backup the same course. So it just make sure the backup do not happen the same time.

Reply:
We could have potential race condition, such as, they are producing the same backup file name, e.g `backup-moodle2-course-2-c1-20250704-1314.mbz`, one of the task may complete first and delete the file. This may interrupt the other task